libsoftwarecontainer/src/gateway/dbus/dbusgateway.cpp

```diff
@@ -18,191 +17,100 @@
  * For further information see LICENSE
  */
 
-
 #include "dbusgateway.h"
-#include "dbusgatewayparser.h"
 
-DBusGateway::DBusGateway(ProxyType type
-                       , const std::string &gatewayDir
-                       , const std::string &name)
-    : Gateway(ID)
-    , m_type(type)
-{
-    m_socket = gatewayDir
-             + (m_type == SessionProxy ? "/sess_" : "/sys_")
-             + name
-             + ".sock";
+DBusGateway::DBusGateway(const std::string &gatewayDir, const std::string &name):
+    Gateway(ID),
+    sessionBus(DBusGatewayInstance::ProxyType::SessionProxy, gatewayDir, name),
+    systemBus(DBusGatewayInstance::ProxyType::SystemProxy, gatewayDir, name)
 
-    m_sessionBusConfig = json_array();
-    m_systemBusConfig = json_array();
+{
 }
 
 DBusGateway::~DBusGateway()
 {
-    if (isActivated()) {
-        teardown();
-    }
 }
 
-
-ReturnCode DBusGateway::readConfigElement(const json_t *element)
+ReturnCode DBusGateway::setConfig(const std::string &config)
 {
-    DBusGatewayParser parser;
+     log_debug() << "Setting config for dbus session bus";
+     ReturnCode sessionBusParseResult = sessionBus.setConfig(config);
 
-    if (isError(parser.parseDBusConfigElement(element, m_sessionBusConfig, m_systemBusConfig))) {
-        log_error() << "Failed to parse DBus configuration element";
-        return ReturnCode::FAILURE;
-    }
+     log_debug() << "Setting config for dbus system bus";
+     ReturnCode systemBusParseResult = systemBus.setConfig(config);
 
-    return ReturnCode::SUCCESS;
-}
+     if (isError(sessionBusParseResult) && isError(systemBusParseResult)) {
+         log_error() << "Neither session nor  system bus could use the given config";
+         return ReturnCode::FAILURE;
+     }
```

Comment:
This would be nice and could be done rather nice without breaking the design. However, it would require the setConfig() to take json_t* as a parameter to not force reimplementation / duplication of the string -> json_t* code.